printer/location_tracker_formatter.ml

```diff
@@ -0,0 +1,226 @@
+      ) t
+
+  let dump ~file t =
+    let ch = open_out (file ^ ".map") in
+    Marshal.to_channel ch t [];
+    close_out ch
+end
+
+module Tracking_formatter = struct
+  module Pending_mapping = struct
+    type t = {
+      label : string option;
+      source : Range.t;
+      start_in_buffer : Location_in_file.t;
+    }
+
+    let create ~label ~source ~start_in_buffer =
+      { label; source; start_in_buffer }
+
+    let to_mapping ~file ~end_in_buffer { label; source; start_in_buffer } =
+      let ir = Range.create ~file ~start:start_in_buffer ~end_:end_in_buffer in
+      Mappings.Item.create ~source ~ir ~label
+  end
+
+  type t =
+    { ppf : Format.formatter
+    ; file : string
+    ; mutable pending_mappings : Pending_mapping.t list
+    ; mutable mappings : Mappings.t
+    ; mutable position : Location_in_file.t
+    }
+
+  let close t = assert (List.length t.pending_mappings = 0)
+  let ppf t = t.ppf
+
+  let track_output_position_of_formatter t =
+    let ppf = t.ppf in
+    let { Format.out_string; out_flush; out_newline; out_spaces; out_indent } =
+      Format.pp_get_formatter_out_functions ppf ()
+    in
+    let track_newline () =
+      (* When a newline is created [line] obviously increases by one, but also
+      now both bol and char are the same as the new line is empty. *)
+      t.position <-
+        { line = t.position.line + 1
+        ; bol = t.position.char
+        ; char = t.position.char
+        }
+    in
+    let track_new_char () =
+      t.position <-
+        { t.position with
+          char = t.position.char + 1
+        }
+    in
+    let out_string str start_pos num_chars =
+      for index = start_pos to start_pos + num_chars - 1 do
+        track_new_char ();
+        if str.[index] = '\n' then track_newline ()
+      done;
+      out_string str start_pos num_chars
+    in
+    let out_functions : Format.formatter_out_functions =
+      { out_string
+      ; out_flush
+      ; (* The default [out_newline] calls [out_string] with a newline
+           character, so we don't need to override [out_newline]. A similar
+           situation exists for [out_spaces] and [out_indent]. *)
+        out_newline
+      ; out_spaces
+      ; out_indent
+      }
+    in
+    Format.pp_set_formatter_out_functions ppf out_functions
+  ;;
+
+  let intercept_location_tags_on_formatter t =
+    let ppf = t.ppf in
+    let stag_functions = Format.pp_get_formatter_stag_functions ppf () in
+    let mark_open_stag = function
+      | Location_mapping_tag (source, label) ->
+        let pending =
+          Pending_mapping.create ~label ~source ~start_in_buffer:t.position
+        in
+        t.pending_mappings <- pending :: t.pending_mappings;
+        ""
+      | stag -> stag_functions.mark_open_stag stag
+    in
+    let mark_close_stag = function
+      | Location_mapping_tag _ ->
+        (match t.pending_mappings with
+         | [] -> failwith "End_pos_tag encountered without matching Start_pos_tag"
```

Comment:
(Will the exception be caught/used by the `Format`
machinery? In other words, should it be a `Misc.fatal`?)

Reply:
The exception will not be caught by the Format machinery -- [t.pending_mappings] is something that we introduced ourselves. We should never be hitting this case because it would represent a bug so I think using failwith is fine in that case.

Reply:
This should be `Misc.fatal_error`, I think.